common/lib/xmodule/xmodule/library_content_module.py

```diff
@@ -320,21 +341,46 @@ def refresh_children(self, request=None, suffix=None):  # pylint: disable=unused
         the version number of the libraries used, so we easily determine if
         this block is up to date or not.
         """
-        lib_tools = self.runtime.service(self, 'library_tools')
-        if not lib_tools:
-            # This error is diagnostic. The user won't see it, but it may be helpful
-            # during debugging.
-            return Response(_(u"Course does not support Library tools."), status=400)
-        user_service = self.runtime.service(self, 'user')
         user_perms = self.runtime.service(self, 'studio_user_permissions')
-        if user_service:
-            # May be None when creating bok choy test fixtures
-            user_id = user_service.get_current_user().opt_attrs.get('edx-platform.user_id', None)
-        else:
-            user_id = None
-        lib_tools.update_children(self, user_id, user_perms)
+        user_id = self.get_user_id()
+        self.tools.update_children(self, user_id, user_perms)
         return Response()
 
+    def studio_post_duplicate(self, store, source_block):
+        """
+        Used by the studio after basic duplication of a source block. We handle the children
+        ourselves, because we have to properly reference the library upstream and set the overrides.
+
+        Otherwise we'll end up losing data on the next refresh.
+        """
+        # The first task will be to refresh our copy of the library to generate the children.
+        # We must do this at the currently set version of the library block. Otherwise we may not have
+        # exactly the same children-- someone may be duplicating an out of date block, after all.
+        user_id = self.get_user_id()
+        user_perms = self.runtime.service(self, 'studio_user_permissions')
+        # pylint: disable=no-member
+        self.tools.update_children(self, user_id, user_perms, version=self.source_library_version)
+
+        # Copy over any overridden settings the course author may have applied to the blocks.
+        def copy_overrides(source, dest):
```

Comment:
(It could instead be a private method that takes all required arguments, which I think would make the code flow easier to follow)